rdtools/availability.py

```diff
@@ -0,0 +1,334 @@
+
+    This function is useful for full system outages when no system data is
+    available at all. However, it does require cumulative production data
+    recorded at the device level and only reports estimated lost production
+    for entire outages rather than timeseries lost power.
+
+    Parameters
+    ----------
+    power : pd.Series
+        Timeseries power data for a system.
+
+    energy : pd.Series
+        Timeseries cumulative energy data for the system. These values
+        must be recorded at the device level (rather than summed by a
+        downstream device like a datalogger or DAS provider) to preserve its
+        integrity across communication interruptions. Units must match
+        ``power`` integrated to hourly energy (e.g. if ``power`` is in kW then
+        ``energy`` must be in kWh).
+
+    subsystem_power : pd.DataFrame
+        Timeseries power data, one column per subsystem. In the typical case,
+        this is inverter AC power data.
+
+    expected_power : pd.Series
+        Expected system power data with the same index as the measured data.
+        This can be modeled from on-site weather measurements if there is no
+        risk of instrument calibration or data gaps. However, because full
+        system outages often cause weather data to be lost as well, it may
+        be more useful to use data from an independent weather station or
+        satellite-based weather provider.
+
+    quantiles : 2-element tuple, default (0.01, 0.99)
+        The quantiles of the error distribution used for the expected energy
+        confidence interval. The lower bound is used to classify outages as
+        either (1) a simple communication interruption with no production loss
+        or (2) a power outage with an associated production loss estimate.
+
+    Returns
+    -------
+    outage_info : pd.DataFrame
+        A dataframe of records about each detected outage, one row per outage.
+        The primary columns of interest are ``type``, which can be either
+        ``'real'`` or ``'comms'`` and reports whether the outage was determined
+        to be a real outage with lost production or just a communications
+        interruption with no production impact; and ``loss`` which reports
+        the estimated production loss for the outage.
+
+    See Also
+    --------
+    rdtools.availability.loss_from_power
+
+    Notes
+    -----
+    This method's ability to detect short-duration outages is limited by the
+    resolution of the system data. For instance, 15-minute averages would not
+    be able to resolve the rapid power cycling of an intermittent inverter.
+    Additionally, the loss at the edges of an outage may be underestimated
+    because of masking by the interval averages.
+
+    The outage detection routine assumes that the input timeseries are
+    continuous, so any data gaps should be filled with nan before using
+    this function.  See :py:func:`rdtools.normalization.interpolate`.
+
```

Comment:
@kanderso-nrel In reference to your response below. I'm not sure I follow the requirement. How would you describe a non-continuous time series?

Reply:
It might be better described as a continuous datetime index.  For example this hourly dataset is discontinuous because it is missing the 01:00 row:
```
2019-01-01 00:00:00    1.0
2019-01-01 02:00:00    3.0
2019-01-01 03:00:00    4.0
```
So to make the index continuous, add a nan row:
```
2019-01-01 00:00:00    1.0
2019-01-01 01:00:00    NaN
2019-01-01 02:00:00    3.0
2019-01-01 03:00:00    4.0
```
It's okay for measured values to be missing -- the issue is only with missing timestamps/rows.

Reply:
I've seen the term "serially complete" used to describe a time series where time points are regular and there are no missing values.

Reply:
Is it required that each value in the index be evenly spaced? If so I might describe it as requiring "a regular index of evenly spaced timestamps" to be explicit about it.

If that is the requirement, then we could check that `index.freq is not None` Although I believe it is possible to have evenly spaced points and `freq==None`, perhaps the most conservative thing to do is raise the error in that case.

Reply:
`index.freq` is not always filled in when an index is created.

```
import pandas as pd
dt = pd.date_range(start='2020-09-22 00:00:00', end='2020-09-22 23:00:00', periods=24)
dt.freq is None
```

Reply:
The NSRDB is frequently advertised as being serially complete.  I don't think we can use it here because even spacing is not a requirement.  This code detects outages by looking at the data values provided to it (numbers or nan), not by trying to find jumps in the index.  The intent is to prevent outages from being overlooked because they were hidden by jumps in the index.  

How about this?

> This function expects outages to be represented in the timeseries by nan, zero, or very low values.  If your DAS does not report data with missing values (so an outage would create a jump in the reported timestamps), you should insert those missing rows before using this function.

Reply:
That makes sense to me, assuming I have understood correctly that your function doesn't assume regular intervals. E.g. It would work with an irregular/uneven time series as long as outages are represented by nan/zero/low

Reply:
It's not a major point, but I think that revision trades one ambiguity for a different one. How are you defining a "jump" in the timestamps if the timestamps are not on regular intervals? And if regular intervals are not required, does the algorithm still work if the timestamps with missing data are simply omitted?

Reply:
I guess the point I'm trying to document is "don't expect the algorithm to find any outages that are shorter than your data intervals, and by the way your data intervals might be way longer than you expect if any rows got dropped during outages".

I'm expecting most users of this code to have a semi-curated dataset from their DAS provider with regular timestamps except when rows get dropped.  In that case, a simple pandas reindex or resample is what is needed before using this code.  For datasets with irregular measurement timing, I see the point that a "jump" is poorly defined, but I'm having trouble coming up with something more precise.  Suggestions are welcome!

Reply:
Revised sentence: "If your DAS does not record data from outages (e.g., a three-hour outage results in three hours of omitted timestamps), you should..."